src/controller_examples/rabbitmq_controller/proof/liveness/liveness.rs

```diff
@@ -68,52 +86,75 @@ proof fn next_with_wf_is_stable()
-    .and(always(lift_state(RMQCluster::busy_disabled())))
+spec fn next_with_wf_and_assumption(rabbitmq: RabbitmqClusterView) -> TempPred<RMQCluster> {
+    always(lift_action(RMQCluster::next()))
+    .and(tla_forall(|input| RMQCluster::kubernetes_api_next().weak_fairness(input)))
+    .and(tla_forall(|input| RMQCluster::controller_next().weak_fairness(input)))
+    .and(tla_forall(|input| RMQCluster::schedule_controller_reconcile().weak_fairness(input)))
+    .and(tla_forall(|input| RMQCluster::builtin_controllers_next().weak_fairness(input)))
+    .and(RMQCluster::disable_crash().weak_fairness(()))
+    .and(RMQCluster::disable_busy().weak_fairness(()))
     .and(always(lift_state(RMQCluster::desired_state_is(rabbitmq))))
-    .and(always(lift_state(RMQCluster::the_object_in_schedule_has_spec_and_uid_as(rabbitmq))))
-    .and(always(lift_state(RMQCluster::the_object_in_reconcile_has_spec_and_uid_as(rabbitmq))))
 }
 
-proof fn assumptions_is_stable(rabbitmq: RabbitmqClusterView)
+spec fn next_with_wf_and_assumption_and_all_invariants(rabbitmq: RabbitmqClusterView) -> TempPred<RMQCluster> {
+    always(lift_action(RMQCluster::next()))
+    .and(tla_forall(|input| RMQCluster::kubernetes_api_next().weak_fairness(input)))
+    .and(tla_forall(|input| RMQCluster::controller_next().weak_fairness(input)))
+    .and(tla_forall(|input| RMQCluster::schedule_controller_reconcile().weak_fairness(input)))
+    .and(tla_forall(|input| RMQCluster::builtin_controllers_next().weak_fairness(input)))
+    .and(RMQCluster::disable_crash().weak_fairness(()))
+    .and(RMQCluster::disable_busy().weak_fairness(()))
+    .and(always(lift_state(RMQCluster::desired_state_is(rabbitmq))))
+    .and(invariants_phase_I(rabbitmq))
+    .and(invariants_phase_II(rabbitmq))
+    .and(invariants_phase_III(rabbitmq))
+    .and(invariants_phase_IV(rabbitmq))
+    .and(invariants_phase_V(rabbitmq))
+    .and(invariants_phase_VI(rabbitmq))
+}
+
+proof fn next_with_wf_and_assumption_is_stable(rabbitmq: RabbitmqClusterView)
     ensures
-        valid(stable(assumptions(rabbitmq))),
+        valid(stable(next_with_wf_and_assumption(rabbitmq))),
 {
-    stable_and_always_n!(
-        lift_state(RMQCluster::crash_disabled()),
-        lift_state(RMQCluster::busy_disabled()),
-        lift_state(RMQCluster::desired_state_is(rabbitmq)),
-        lift_state(RMQCluster::the_object_in_schedule_has_spec_and_uid_as(rabbitmq)),
-        lift_state(RMQCluster::the_object_in_reconcile_has_spec_and_uid_as(rabbitmq))
+    always_p_is_stable(lift_state(RMQCluster::desired_state_is(rabbitmq)));
+    next_with_wf_is_stable();
+    stable_and_temp(
+        next_with_wf(),
+        always(lift_state(RMQCluster::desired_state_is(rabbitmq)))
     );
 }
 
 // The safety invariants that are required to prove liveness.
-spec fn invariants(rabbitmq: RabbitmqClusterView) -> TempPred<RMQCluster> {
+spec fn invariants_phase_I(rabbitmq: RabbitmqClusterView) -> TempPred<RMQCluster> {
```

Comment:
Could you make it explicit that these invariants hold from the init point?

Reply:
Actually I want to say Phase 0 but there is no zero in roman numerals 👎

Reply:
You can say `phase_zero` if you don't mind the inconsistency